In [56]:
import os
import pandas as pd

In [57]:
main_task_path = os.path.join('data', 'main_task.csv')

In [58]:
df = pd.read_csv(main_task_path)

In [59]:
df.head()

Restaurant_id       City                                      Cuisine Style  \
0       id_5569      Paris            ['European', 'French', 'International']   
1       id_1535  Stockholm                                                NaN   
2        id_352     London  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...   
3       id_3456     Berlin                                                NaN   
4        id_615     Munich  ['German', 'Central European', 'Vegetarian Fri...   

   Ranking  Rating Price Range  Number of Reviews  \
0   5570.0     3.5    $$ - $$$              194.0   
1   1537.0     4.0         NaN               10.0   
2    353.0     4.5        $$$$              688.0   
3   3458.0     5.0         NaN                3.0   
4    621.0     4.0    $$ - $$$               84.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   

                                              URL_TA     ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...  d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...  d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...  d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...  d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...  d6864963

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  object 
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  37457 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), object(7)
memory usage: 3.1+ MB


In [5]:
# Превратим данные в столбце Restaurant_id в числовые данные
df['Restaurant_id'] = df['Restaurant_id'].apply(lambda id: int(id[3:]))

In [20]:
df.head(10)

Restaurant_id        City  \
0           5569       Paris   
1           1535   Stockholm   
2            352      London   
3           3456      Berlin   
4            615      Munich   
5           1418      Oporto   
6           1720       Milan   
7            825  Bratislava   
8           2690      Vienna   
9           4209        Rome   

                                       Cuisine Style  Ranking  Rating  \
0            ['European', 'French', 'International']   5570.0     3.5   
1                                                NaN   1537.0     4.0   
2  ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...    353.0     4.5   
3                                                NaN   3458.0     5.0   
4  ['German', 'Central European', 'Vegetarian Fri...    621.0     4.0   
5                                                NaN   1419.0     3.0   
6                               ['Italian', 'Pizza']   1722.0     4.0   
7                                        ['Italian']    826.0     3.0   
8                                                NaN   2692.0     4.0   
9                  ['Italian', 'Pizza', 'Fast Food']   4210.0     4.0   

  Price Range  Number of Reviews  \
0    $$ - $$$              194.0   
1         NaN               10.0   
2        $$$$              688.0   
3         NaN                3.0   
4    $$ - $$$               84.0   
5         NaN                2.0   
6           $               50.0   
7         NaN                9.0   
8         NaN                NaN   
9           $               55.0   

                                             Reviews  \
0  [['Good food at your doorstep', 'A good hotel ...   
1  [['Unique cuisine', 'Delicious Nepalese food']...   
2  [['Catch up with friends', 'Not exceptional'],...   
3                                           [[], []]   
4  [['Best place to try a Bavarian food', 'Nice b...   
5  [['There are better 3 star hotel bars', 'Amazi...   
6  [['Excellent simple local eatery.', 'Excellent...   
7  [['Wasting of money', 'excellent cuisine'], ['...   
8                                           [[], []]   
9  [['Clean efficient staff', 'Nice little pizza ...   

                                              URL_TA      ID_TA  
0  /Restaurant_Review-g187147-d1912643-Reviews-R_...   d1912643  
1  /Restaurant_Review-g189852-d7992032-Reviews-Bu...   d7992032  
2  /Restaurant_Review-g186338-d8632781-Reviews-RO...   d8632781  
3  /Restaurant_Review-g187323-d1358776-Reviews-Es...   d1358776  
4  /Restaurant_Review-g187309-d6864963-Reviews-Au...   d6864963  
5  /Restaurant_Review-g189180-d12503536-Reviews-D...  d12503536  
6  /Restaurant_Review-g187849-d5808504-Reviews-Pi...   d5808504  
7  /Restaurant_Review-g274924-d3199765-Reviews-Ri...   d3199765  
8  /Restaurant_Review-g190454-d12845029-Reviews-G...  d12845029  
9  /Restaurant_Review-g187791-d8020681-Reviews-Qu...   d8020681

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40000 entries, 0 to 39999
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Restaurant_id      40000 non-null  int64  
 1   City               40000 non-null  object 
 2   Cuisine Style      30717 non-null  object 
 3   Ranking            40000 non-null  float64
 4   Rating             40000 non-null  float64
 5   Price Range        26114 non-null  object 
 6   Number of Reviews  37457 non-null  float64
 7   Reviews            40000 non-null  object 
 8   URL_TA             40000 non-null  object 
 9   ID_TA              40000 non-null  object 
dtypes: float64(3), int64(1), object(6)
memory usage: 3.1+ MB


### 1.3.1 Вопросы о ценах

#### Сколько вариантов непустых значений встречается в столбце 'Price Range'?

In [8]:
price_ranges = pd.Series(df[~df['Price Range'].isna()]['Price Range'])
display(price_ranges.unique())

array(['$$ - $$$', '$$$$', '$'], dtype=object)

#### Ответ: 3

#### Как в датафрейме обозначается самый низкий уровень цен?

#### Ответ: можно догадаться, что '$'

#### Как в датафрейме обозначается самый высокий уровень цен?

#### Ответ: можно догадаться, что \\$\\$\\$\\$

#### Сколько ресторанов относятся к среднему ценовому сегменту?

In [9]:
display(price_ranges.value_counts()[0])

18412

#### Ответ: 18412

### 1.3.2 Вопросы о городах 

#### Сколько городов представлено в наборе данных?

In [10]:
df['City'].nunique()

31

#### Ответ: 31

#### Сколько ресторанов расположены в столичных городах?

In [ ]:
# Формируем множество столиц из датасета, найденного на data.world

In [11]:
import json as js

In [12]:
with open('data/country-by-capital-city.json') as json_file:
    country_capital_city_dict = js.load(json_file)

In [16]:
capitals = set()
for row in country_capital_city_dict:
    capitals.add(row['city'])

In [33]:
# Добавим функцию определяющую входит ли город в множество столиц
def is_capital(city):
    return city in capitals

# Применим эту функцию к столбцу 'City'и суммируем результат
df[df['City'].apply(is_capital)]['Restaurant_id'].nunique()

11242

In [19]:
df[df['City'].apply(is_capital)]['City']

0             Paris
1         Stockholm
2            London
3            Berlin
7        Bratislava
            ...    
39992     Stockholm
39993         Paris
39994         Paris
39996         Paris
39997     Stockholm
Name: City, Length: 21436, dtype: object

In [26]:
df.groupby('Restaurant_id')['Restaurant_id'].count().head(20)

Restaurant_id
0     15
1      9
2     14
3      8
4     13
5     11
6      8
7     11
8     13
9     12
10    14
11     8
12    14
13    11
14    15
15     7
16     8
17    14
18    10
19    13
Name: Restaurant_id, dtype: int64

In [28]:
df[df['Restaurant_id'] == 0]['City']

4453       Hamburg
5441          Rome
5936        London
6194         Milan
7199        Madrid
12734         Oslo
12803    Stockholm
19744       Krakow
21372         Lyon
21805       Lisbon
22863    Edinburgh
23201       Vienna
23892       Warsaw
24120    Amsterdam
34458     Budapest
Name: City, dtype: object

In [23]:
sum(df['Restaurant_id'] == 0)

15

#### Ответ: 21436 - неправильный!!! Причины: был выбран недостаточно информативный датасет по городам. В этом датасете присутствовали только первичные столицы, а административные столицы отсутствовали. Вывод: использовать правильный датасет.

#### Благодаря члену группы sfdatascience1.slack.com с ником Mikhail решение проблемы было найдено в использовании датасета https://www.kaggle.com/viswanathanc/world-cities-datasets#worldcities.csv и в расширении понятия 'столица' кроме тех городов в поле 'capital', которых стоит значение 'primary' еще и городом 'Edinburgh'. Весьма странно, но куда деваться??? Еще раз огромное спасибо Mikhail за титаническую работу, которую он проделал.

In [34]:
cities = pd.read_csv('data/worldcities.csv')

In [37]:
cities[cities['city'] == 'Edinburgh']

city city_ascii      lat     lng         country iso2 iso3  \
1036  Edinburgh  Edinburgh  55.9483 -3.2191  United Kingdom   GB  GBR   

              admin_name capital  population          id  
1036  Edinburgh, City of   admin    504966.0  1826492520

In [38]:
cities.head()

city   city_ascii      lat       lng        country iso2 iso3  \
0        Tokyo        Tokyo  35.6850  139.7514          Japan   JP  JPN   
1     New York     New York  40.6943  -73.9249  United States   US  USA   
2  Mexico City  Mexico City  19.4424  -99.1310         Mexico   MX  MEX   
3       Mumbai       Mumbai  19.0170   72.8570          India   IN  IND   
4    São Paulo    Sao Paulo -23.5587  -46.6250         Brazil   BR  BRA   

         admin_name  capital  population          id  
0             Tōkyō  primary  35676000.0  1392685764  
1          New York      NaN  19354922.0  1840034016  
2  Ciudad de México  primary  19028000.0  1484247881  
3       Mahārāshtra    admin  18978000.0  1356226629  
4         São Paulo    admin  18845000.0  1076532519

In [39]:
cities['capital'].unique()

array(['primary', nan, 'admin', 'minor'], dtype=object)

#### Как видно в датафрейме cities не столицами считаются только те города, поле capital которых не заполнено никакими данными

In [50]:
# Вычистим города с отсутствующими данными в поле 'capital'
capitals_df = cities.dropna(subset=['capital'])

# Сформируем множество столиц всех типов, кроме типа 'minor'
capitals = set(capitals_df[capitals_df['capital'] == 'primary']['city'])

In [51]:
capitals

{'Abidjan',
 'Abu Dhabi',
 'Abuja',
 'Accra',
 'Addis Ababa',
 'Al Quds',
 'Algiers',
 'Amman',
 'Amsterdam',
 'Andorra la Vella',
 'Ankara',
 'Antananarivo',
 'Apia',
 'Ashgabat',
 'Asmara',
 'Asunción',
 'Athens',
 'Baghdad',
 'Baku',
 'Bamako',
 'Bandar Seri Begawan',
 'Bangkok',
 'Bangui',
 'Banjul',
 'Basseterre',
 'Beijing',
 'Beirut',
 'Belgrade',
 'Belmopan',
 'Berlin',
 'Bern',
 'Bishkek',
 'Bissau',
 'Bloemfontein',
 'Bogotá',
 'Brasília',
 'Bratislava',
 'Brazzaville',
 'Bridgetown',
 'Brussels',
 'Bucharest',
 'Budapest',
 'Buenos Aires',
 'Bujumbura',
 'Cairo',
 'Canberra',
 'Cape Town',
 'Capitol Hill',
 'Caracas',
 'Castries',
 'Chisinau',
 'Colombo',
 'Conakry',
 'Copenhagen',
 'Cotonou',
 'Dakar',
 'Damascus',
 'Dar es Salaam',
 'Dhaka',
 'Dili',
 'Djibouti',
 'Dodoma',
 'Doha',
 'Dublin',
 'Dushanbe',
 'Freetown',
 'Funafuti',
 'Gaborone',
 'Georgetown',
 'Gitega',
 'Guatemala City',
 'Hagåtña',
 'Hanoi',
 'Harare',
 'Havana',
 'Helsinki',
 'Honiara',
 'Islamabad',
 '

In [53]:
capitals.add('Edinburgh')

In [55]:
# Применим эту функцию к столбцу 'City'и суммируем результат
df[df['City'].apply(is_capital)]

Restaurant_id        City  \
0               5569       Paris   
1               1535   Stockholm   
2                352      London   
3               3456      Berlin   
7                825  Bratislava   
...              ...         ...   
39993           6057       Paris   
39994           7074       Paris   
39996           6340       Paris   
39997           1649   Stockholm   
39998            640      Warsaw   

                                           Cuisine Style  Ranking  Rating  \
0                ['European', 'French', 'International']   5570.0     3.5   
1                                                    NaN   1537.0     4.0   
2      ['Japanese', 'Sushi', 'Asian', 'Grill', 'Veget...    353.0     4.5   
3                                                    NaN   3458.0     5.0   
7                                            ['Italian']    826.0     3.0   
...                                                  ...      ...     ...   
39993                                                NaN   6058.0     4.0   
39994  ['French', 'Fast Food', 'Mediterranean', 'Heal...   7075.0     4.5   
39996  ['French', 'American', 'Bar', 'European', 'Veg...   6341.0     3.5   
39997                              ['Japanese', 'Sushi']   1652.0     4.5   
39998  ['Polish', 'European', 'Eastern European', 'Ce...    641.0     4.0   

      Price Range  Number of Reviews  \
0        $$ - $$$              194.0   
1             NaN               10.0   
2            $$$$              688.0   
3             NaN                3.0   
7             NaN                9.0   
...           ...                ...   
39993         NaN               16.0   
39994    $$ - $$$                6.0   
39996    $$ - $$$              542.0   
39997         NaN                4.0   
39998    $$ - $$$               70.0   

                                                 Reviews  \
0      [['Good food at your doorstep', 'A good hotel ...   
1      [['Unique cuisine', 'Delicious Nepalese food']...   
2      [['Catch up with friends', 'Not exceptional'],...   
3                                               [[], []]   
7      [['Wasting of money', 'excellent cuisine'], ['...   
...                                                  ...   
39993  [['Naught but VERY nice', 'Yum'], ['01/23/2017...   
39994                                           [[], []]   
39996  [['Parisian atmosphere', 'Bit pricey but inter...   
39997  [['Good by swedish standards', 'A hidden jewel...   
39998  [['Underground restaurant', 'Oldest Restaurant...   

                                                  URL_TA      ID_TA  
0      /Restaurant_Review-g187147-d1912643-Reviews-R_...   d1912643  
1      /Restaurant_Review-g189852-d7992032-Reviews-Bu...   d7992032  
2      /Restaurant_Review-g186338-d8632781-Reviews-RO...   d8632781  
3      /Restaurant_Review-g187323-d1358776-Reviews-Es...   d1358776  
7      /Restaurant_Review-g274924-d3199765-Reviews-Ri...   d3199765  
...                                                  ...        ...  
39993  /Restaurant_Review-g187147-d10532509-Reviews-B...  d10532509  
39994  /Restaurant_Review-g187147-d12621988-Reviews-C...  d12621988  
39996  /Restaurant_Review-g187147-d1800036-Reviews-La...   d1800036  
39997  /Restaurant_Review-g189852-d947615-Reviews-Sus...    d947615  
39998  /Restaurant_Review-g274856-d1100838-Reviews-Ho...   d1100838  

[30424 rows x 10 columns]

#### Ответ: 30424, блин!!!